sharkle@lufinkey/files/sharkle@lufinkey/desklet.js

```diff
@@ -0,0 +1,285 @@
+
+const Desklet = imports.ui.desklet;
+const Settings = imports.ui.settings;
+const St = imports.gi.St;
+const Clutter = imports.gi.Clutter;
+const Lang = imports.lang;
+const Mainloop = imports.mainloop;
+const Sound = imports.ui.soundManager;
+
+
+
+function Animation(fps)
+{
+	this._init(fps);
+}
+Animation.prototype = {
+	_init: function(fps)
+	{
+		this.images = [];
+		this.fps = fps;
+	},
+	
+	addImage: function(image)
+	{
+		this.images.push(image);
+	},
+}
+
+
+
+function AnimationPlayer()
+{
+	this._init();
+}
+AnimationPlayer.prototype = {
+	_init: function(){
+		this.loop_id = 0;
+		this.animation = null;
+		this.frame = 0;
+		this.onUpdate = null;
+		this.onFinish = null;
+	},
+	
+	setAnimation: function(animation)
+	{
+		if(this.animation!=null)
+		{
+			Mainloop.source_remove(this.loop_id);
+			this.loop_id = 0;
+		}
+		this.frame = 0;
+		this.animation = animation;
+		if(this.animation!=null)
+		{
+			this.loop_id = Mainloop.timeout_add(1000.0/this.animation.fps, Lang.bind(this, this._update_loop));
```

Comment:
When removing the desklet the looping doesn't end. These errors are seen continuously, note these can only be seen using cjs master:

```
(cinnamon:18866): Cjs-CRITICAL **: 02:45:00.794: Object St.Bin (0x55a3a4d4a3b0), has been already deallocated - impossible to access to it. This might be caused by the fact that the object has been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs
== Stack trace for context 0x55a3a2cc64c0 ==
#0 0x55a3a3604c48 i   /usr/share/cinnamon/js/misc/fileUtils.js line 210 > Function:107 (0x7f92437694d8 @ 30)
#1 0x55a3a3604b90 i   /usr/share/cinnamon/js/misc/fileUtils.js line 210 > Function:91 (0x7f92437699a0 @ 237)
#2 0x55a3a3604ad0 i   self-hosted:914 (0x7f92ddee3ef0 @ 346)

(cinnamon:18866): Cjs-CRITICAL **: 02:45:00.794: Object St.Bin (0x55a3a4d4a3b0), has been already deallocated - impossible to access to it. This might be caused by the fact that the object has been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs
== Stack trace for context 0x55a3a2cc64c0 ==
#0 0x55a3a3604c48 i   /usr/share/cinnamon/js/misc/fileUtils.js line 210 > Function:112 (0x7f92437694d8 @ 86)
#1 0x55a3a3604b90 i   /usr/share/cinnamon/js/misc/fileUtils.js line 210 > Function:91 (0x7f92437699a0 @ 237)
#2 0x55a3a3604ad0 i   self-hosted:914 (0x7f92ddee3ef0 @ 346)

```

You could use a function [like this](https://github.com/linuxmint/cinnamon-spices-applets/blob/64c6130057a2ae2dce4e0fcc4b663e997ae8b3e3/IcingTaskManager%40json/files/IcingTaskManager%40json/3.8/utils.js#L189) to check if an object is finalized, but stopping the loop in `on_desklet_removed` should be enough.